File: Easy/Arrays_Strings/14_Longest_Common_Prefix.c
/*
Problem: 14. Longest Common Prefix
Approach:
- String manipulation using character arrays

Time Complexity: O(n * m)
- n is the number of strings
- m is the length of the prefix
Space Complexity: O(1)
*/

#include <string.h>
#include <stddef.h> // For size_t

char* longestCommonPrefix(char** strs, int strsSize) {
    if (strsSize == 0) {
        return "";
    }
    
    char *LCP = strs[0];
    for (int i = 1; i < strsSize; i++) {
        char *current_str = strs[i];
        
        while (strncmp(current_str, LCP, strlen(LCP)) != 0) {
            size_t current_len = strlen(LCP);
            if (current_len == 0) {
                return "";
            }
            LCP[current_len - 1] = '\0';
        }
    }
    return LCP;
}

File: Easy/Arrays_Strings/14_Longest_Common_Prefix.py
"""
Problem: Longest Common Prefix

Approach:
- Assume the first string is the common prefix
- Compare it with each string
- While the current string does not start with the prefix, remove the last character from the prefix
- Return the remaining prefix

Time Complexity: O(n * m)
Space Complexity: O(1)

where:
n = number of strings
m = length of the shortest string
"""

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        prefix = strs[0]
    
        for word in strs:
            while not word.startswith(prefix):
                prefix = prefix[:-1]
        
        return prefix
